### PR TITLE
feat: Upgrade Hugr and start using the shared Pydantic model

### LINKED ISSUE #198: Upgrade to Hugr v0.4
Blocked by
* https://github.com/CQCL/hugr/issues/935
* https://github.com/CQCL/hugr/issues/941

### PR DESCRIPTION
**Pull request by mark-koch**
This PR makes Guppy depend on the Hugr 0.4 release candidate `0.4.0-alpha.1`. Closes #198

Most of the edits are related to changes in the Pydantic model. However, we also had to get rid of `TypeApply` since this no longer exists in Hugr. Therefore, Guppy now rejects all programs that use polymorphic functions as values if the type arguments cannot be inferred

### CODE REVIEW

File: guppylang/compiler/cfg_compiler.py [hunk 1]
```diff
@@ -92,7 +91,11 @@ def compile_bb(
                 graph=graph,
                 unit_sum=branch_port,
                 output_vars=[
-                    [v for v in row if not v.ty.linear or is_return_var(v.name)]
+                    [
+                        v
+                        for v in sort_vars(row)
+                        if not v.ty.linear or is_return_var(v.name)
+                    ]
```
Comment (mark-koch) [inline guppylang/compiler/cfg_compiler.py:98]: I moved the call to `sort_vars` from `choose_vars_for_tuple_sum()` below to here as a drive-by since this is more consistent with the other sorting logic in this method

File: guppylang/compiler/expr_compiler.py [hunk 2]
```diff
@@ -323,15 +341,13 @@ def python_value_to_hugr(v: Any, exp_ty: Type) -> val.Value | None:
                 python_value_to_hugr(elt, ty)
                 for elt, ty in zip(elts, exp_ty.element_types)
             ]
-            if any(value is None for value in vs):
-                return None
-            return val.Tuple(vs=vs)
+            if doesnt_contain_none(vs):
+                return ops.Value(ops.TupleValue(vs=vs))
```
Comment (mark-koch) [inline guppylang/compiler/expr_compiler.py:345]: Mypy started complaining about this, so I added a `doesnt_contain_none` type guard

File: tests/integration/test_poly.py [hunk 3]
```diff
@@ -259,6 +261,7 @@ def main() -> None:
     validate(module.compile())
 
 
+@pytest.mark.skip("Not yet supported")
```
Comment (mark-koch) [inline tests/integration/test_poly.py:264]: This worked before, but can no longer compile this since `TypeApply` is gone. We can recover this test case by having smarter type arg inference in the future

File: tests/integration/test_py.py [hunk 4]
```diff
@@ -142,6 +142,7 @@ def foo(q1: qubit, q2: qubit, q3: qubit) -> tuple[qubit, qubit, qubit]:
     validate(module.compile())
 
 
+@pytest.mark.skip("Requires Tket2 upgrade")
```
Comment (mark-koch) [inline tests/integration/test_py.py:145]: Pytket measurements in tket2 now returns a custom `Bit` type instead of the prelude bool. This means, we need to load the Tket1 extension for validation, but the extension is still at Hugr v0.3

File: guppylang/tys/ty.py [hunk 5]
```diff
@@ -269,9 +270,20 @@ def parametrized(self) -> bool:
 
     def to_hugr(self) -> tys.Type:
         """Computes the Hugr representation of the type."""
+        if self.parametrized:
+            raise InternalGuppyError(
+                "Tried to convert parametrised function type to Hugr. Use "
+                "`to_hugr_poly` instead"
+            )
+        ins = [t.to_hugr() for t in self.inputs]
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply guppylang/tys/ty.py:278]: looks good tho 👍

File: guppylang/tys/ty.py [hunk 6]
```diff
@@ -376,12 +390,13 @@ def intrinsically_linear(self) -> bool:
...
     def to_hugr(self) -> tys.Type:
         """Computes the Hugr representation of the type."""
-        if all(
-            isinstance(e, TupleType) and len(e.element_types) == 0
-            for e in self.element_types
-        ):
-            return tys.UnitSum(size=len(self.element_types))
-        return tys.GeneralSum(row=[t.to_hugr() for t in self.element_types])
+        rows = [type_to_row(ty) for ty in self.element_types]
+        if all(len(row) == 0 for row in rows):
```
Comment (acl-cqc) [inline guppylang/tys/ty.py:394]: Consider
```
sum_inner = tys.UnitSum(size=len(rows)) if all(len(row)==0 for row in rows) else tys.GeneralSum(rows=[[ty.to_hugr() for ty in row] for row in rows])
return tys.Type(tys.TaggedSumType(st=tys.SumType(sum_inner)))
```

Also the `[[ty.to_hugr() for ty in row] for row in rows]` could be a useful utility method, there are quite a few such double-traversals. `sum_variants_to_hugr`, say

File: tests/error/iter_errors/end_missing.py [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
-from guppylang.hugr import tys
 from guppylang.module import GuppyModule
+from guppylang.tys.ty import NoneType
 
 
 module = GuppyModule("test")
 
 
-@guppy.type(module, tys.TupleType(inner=[]))
+@guppy.type(module, NoneType().to_hugr())
```
Comment (acl-cqc) [inline tests/error/iter_errors/end_missing.py:9]: hmmm, perhaps we should have (somewhere outside test)
```
HUGR_UNIT_TYPE = TupleType([]).to_hugr()
```
and then `NoneType.to_hugr()` just returns that - might possibly be a bit clearer

**Review by acl-cqc — APPROVED** [review]
Massive PR but generally looks good, thanks Mark :). A lot of small comments and questions and suggestions for future....

File: guppylang/tys/builtin.py [hunk 8]
```diff
@@ -95,15 +96,16 @@ def check_instantiate(
         return super().check_instantiate(args, loc)
 
 
-def _list_to_hugr(args: Sequence[Argument]) -> tys.Opaque:
-    return tys.Opaque(
+def _list_to_hugr(args: Sequence[Argument]) -> tys.Type:
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/tys/builtin.py:99]: This function is used as the `to_hugr` hook of the list type:

https://github.com/CQCL/guppylang/blob/11a1619ef75da82584cd7bddb467e8c19a21aae8/guppylang/tys/builtin.py#L130-L137

`_ListTypeDef` is an `OpaqueTypeDef` where the `to_hugr` function needs to have this signature:

https://github.com/CQCL/guppylang/blob/11a1619ef75da82584cd7bddb467e8c19a21aae8/guppylang/definition/ty.py#L32-L38

File: guppylang/tys/const.py [hunk 9]
```diff
@@ -36,7 +37,7 @@ class ConstValue(Const):
 
     # Hugr encoding of the value
     # TODO: We might need a Guppy representation of this...
-    value: val.Value
+    value: ops.Const
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/tys/const.py:40]: Yeah, the Pydantic model in the Hugr repo moved values into the `ops` module

File: scripts/generate_schema.py [hunk 10]
```diff
@@ -6,10 +6,10 @@
...
 
-from guppylang.hugr.raw import RawHugr
+from hugr.serialization.serial_hugr import SerialHugr
 
 if __name__ == "__main__":
     out_dir = Path(sys.argv[-1])
 
     with (out_dir / "hugr_schema_v0.json").open("w") as f:
-        json.dump(TypeAdapter(RawHugr).json_schema(), f)
+        json.dump(TypeAdapter(SerialHugr).json_schema(), f)
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply scripts/generate_schema.py:15]: Oh this script actually isn't needed anymore since schema generation now lives in the Hugr repo. I'll delete the file

File: guppylang/definition/custom.py [hunk 11]
```diff
@@ -138,12 +144,7 @@ def load(
                 "This function does not support usage in a higher-order context",
                 node,
             )
-
-        if self.ty.parametrized:
-            raise InternalGuppyError(
-                "Can't yet generate higher-order versions of custom functions. This "
-                "requires generic function *definitions*"
-            )
+        assert len(self.ty.params) == len(type_args)
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/definition/custom.py:147]: There are non-polymorphic custum Guppy functions whose type cannot be represented in the type system. One example are functions that are implemented via dunder methods (e.g. `iter(x)`, `len(x)`, ...). Therefore, these need `self.higher_order_value=True` which cannot be inferred from the type.

[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/definition/custom.py:147]: > Does RawCustomFunctionDef actually need compile_call or could that be moved into CustomFuncDef?

This is a hack to allow stuff like

https://github.com/CQCL/guppylang/blob/11a1619ef75da82584cd7bddb467e8c19a21aae8/guppylang/prelude/_internal.py#L257-L274

since `Int.__float__` etc are `RawCustomFunctionDef`s. I agree this is not nice but hopefully only a temporary hack. Once we have proper linking, we can define `truediv` etc via actual Guppy function defs and no longer need these `CustomCallCompiler`s.

File: guppylang/hugr_builder/hugr.py [hunk 12]
```diff
@@ -180,9 +191,12 @@ def update_op(self) -> None:
         Feeds type information from the in- and out-ports to the operation class to
         update signature information. This function must be called before serialisation.
         """
-        in_types = [t.to_hugr() for t in self.in_port_types]
-        out_types = [t.to_hugr() for t in self.out_port_types]
-        self.op.insert_port_types(in_types, out_types)
+        if isinstance(self.op.root, ops.BaseOp) and not isinstance(
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:194]: The problem is not the call to `insert_port_types` but the call to `t.to_hugr()` since for these ops, `t` could be a parametrised `PolyFuncType` which can't be embedded into a type. I'll add a comment explaining what's going.

The proper way to deal with this issue would be adding a `Function` edge kind but this would have needed a big refactor of how Guppys internal Python Hugr data structure works. Since we're going to write a proper Python Hugr builder anyways, I thought it's probably not worth the effort.

File: guppylang/hugr_builder/hugr.py [hunk 13]
```diff
@@ -684,35 +748,35 @@ def insert_order_edges(self) -> "Hugr":
         This action must be performed before serialisation.
         """
         for n in self.nodes():
-            if isinstance(n.op, ops.DataflowOp) and isinstance(
+            if isinstance(n.op.root, ops.DataflowOp) and isinstance(
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:751]: Oh nice I was not aware of this! We need to keep adding order edges for non-local edges though

File: guppylang/hugr_builder/hugr.py [hunk 14]
```diff
@@ -746,16 +810,16 @@ def to_raw(self) -> raw.RawHugr:
...
         for n in all_nodes:
-            if isinstance(n.op, ops.Input):
+            if isinstance(n.op.root, ops.Input):
                 input_nodes.append(n)
-            elif isinstance(n.op, ops.Output):
+            elif isinstance(n.op.root, ops.Output):
                 output_nodes.append(n)
             elif (
-                isinstance(n.op, ops.DataflowBlock)
+                isinstance(n.op.root, ops.DataflowBlock)
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:818]: > Is this to put the entry block in first?

Yes!

> Possibly you avoid generating such an entry block yourself, in which case this is probably worth a comment

Yeah, Guppy will never generate jumps to the entry BB since loops always create a new BB for the test and jump back to that. I'll add a comment.

> What if there's an infinite loop, so the code after it is unreachable (and really does have no edges in) - can we generate such a CFG?

We can't generate these right now. Guppy doesn't know that the statement `True` is always true, so it will always generate a branch for the case that it's false. But once Guppy is smarter about this, it will indeed become a problem! I'll move the case checking for exits above the one checking for entries, this should fix the issue.

File: tests/integration/test_poly.py (hunk 3 above)
Comment (mark-koch) [reply tests/integration/test_poly.py:264]: Exactly!

File: guppylang/hugr_builder/hugr.py [hunk 15]
```diff
@@ -262,6 +277,8 @@ class BlockNode(DFContainingNode, CFNode):
 OrderEdge = tuple["Node", "Node"]
 ORDER_EDGE_KEY = (-1, -1)
 
+UNDEFINED: ops.NodeID = -1
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:280]: > I reckon we should think about changing the python wrappers to provide a default (empty/zero/etc.) value for `BaseOp.parent`...

That's how it was before, but the Pydantic model in the Hugr repo got rid of it to have a more accurate schema :(

File: guppylang/hugr_builder/hugr.py [hunk 16]
```diff
@@ -68,6 +69,17 @@ class OutPortCF(OutPort):
     """A control-flow output port."""
 
 
+class DummyOp(OpType):
+    """A dummy Hugr op that is replaced with an extern call during serialisation."""
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:73]: Yeah, the docstring is misleading, I'll update it 👍

File: guppylang/hugr_builder/hugr.py [hunk 17]
```diff
@@ -437,65 +464,112 @@ def add_conditional(
...
+        self,
+        variants: Sequence[TypeRow],
+        tag: int,
+        inputs: list[OutPortV],
+        parent: Node | None = None,
     ) -> VNode:
         """Adds a `Tag` node to the graph."""
-        types = [ty.to_hugr() for ty in variants]
-        assert inp.ty == variants[tag]
+        types = [[ty.to_hugr() for ty in row] for row in variants]
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:513]: I went with `hugr_variants`

File: guppylang/hugr_builder/hugr.py [hunk 18]
```diff
@@ -528,37 +602,20 @@ def add_partial(
...
-            inputs=[func_port],
-            output_types=[result_ty],
-            parent=parent,
+            DummyOp("partial"), None, [new_ty], parent, [*inputs, def_port]
         )
 
     def add_def(
         self, fun_ty: FunctionType, parent: Node | None, name: str
     ) -> DFContainingVNode:
         """Adds a `Def` node to the graph."""
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:611]: Probably a relic from the past :D

File: guppylang/hugr_builder/hugr.py [hunk 19]
```diff
@@ -662,7 +719,7 @@ def remove_dummy_nodes(self) -> "Hugr":
             raise ValueError("Dummy node removal requires a module root node")
         used_names: dict[str, int] = {}
         for n in list(self.nodes()):
-            if isinstance(n, VNode) and isinstance(n.op, ops.DummyOp):
+            if isinstance(n, VNode) and isinstance(n.op, DummyOp):
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:722]: The problem with the library is that we don't have Hugr linking yet, so imo not really worth putting effort into this yet.

>  transferring the type arguments onto each Hugr Call, as you'll need to eventually anyway

Indeed, the place to do that will be the `OpCompiler` in `custom.py`, then we won't have to generate Call but can directly place the op with the right args.

File: guppylang/hugr_builder/hugr.py [hunk 20]
```diff
@@ -792,7 +858,7 @@ def to_raw(self) -> raw.RawHugr:
 
         if self.root is None:
             raise ValueError("Raw Hugr requires a root node")
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/hugr.py:860]: Nice spot, this doesn't look right!

File: guppylang/hugr_builder/visualise.py [hunk 21]
```diff
@@ -63,13 +63,8 @@
...
     <TD>
       <TABLE BORDER="0" CELLBORDER="0">
-        <TR>
-          <TD>
-            <FONT POINT-SIZE="{fontsize}" FACE="{fontface}" COLOR="{label_color}">
-              <B>{node_label}</B>{node_data}
-            </FONT>
-          </TD>
-        </TR>
+        <TR><TD><FONT POINT-SIZE="{fontsize}" FACE="{fontface}"
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/hugr_builder/visualise.py:66]: For some reason, the spacing in the visualisation started to look off. Chaning the indentation fixed it

File: guppylang/prelude/builtins.py [hunk 22]
```diff
@@ -34,7 +36,7 @@
 
 @guppy.extend_type(builtins, bool_type_def)
 class Bool:
-    @guppy.hugr_op(builtins, logic_op("And", [tys.BoundedNatArg(n=2)]))
+    @guppy.hugr_op(builtins, logic_op("And", [tys.TypeArg(tys.BoundedNatArg(n=2))]))
```
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/prelude/builtins.py:39]: Hugr has an n-ary `And` op. we need to pass 2 to get the binary and

File: tests/error/iter_errors/end_missing.py (hunk 7 above)
Comment (mark-koch) [reply tests/error/iter_errors/end_missing.py:9]: Hmm not sure if this is clearer. In `TupleType` we're now doing a similar thing by calling `to_hugr()` on a `SumType` constructed in-place. But I don't really feel strongly about this.

Btw since types are immutable, we could drive this further and just intern all types. But this sounds like premature optimisation

File: guppylang/tys/builtin.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/tys/builtin.py:99]: However, I added an assert to ensure that it's a single argument of kind type

File: tests/integration/test_py.py (hunk 4 above)
Comment (mark-koch) [reply tests/integration/test_py.py]: Tket2 is on the Hugr 0.4 release candidate now, but it still doesn't work. We need to capture the `LBit` type or write a conversion pass that turns TKET1 measurements into TKET2 measurements...

File: guppylang/prelude/builtins.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply guppylang/prelude/builtins.py:39]: oh, gawd, right, sorry!

File: guppylang/hugr_builder/hugr.py [hunk 23]
```diff
@@ -746,23 +794,29 @@ def to_raw(self) -> raw.RawHugr:
...
         all_nodes = self.nodes()
-        root_node = next(all_nodes)
         for n in all_nodes:
-            if isinstance(n.op, ops.Input):
+            if n is self.root:
+                pass
+            elif isinstance(n.op.root, ops.Input):
                 input_nodes.append(n)
-            elif isinstance(n.op, ops.Output):
+            elif isinstance(n.op.root, ops.Output):
```
Comment (acl-cqc) [inline guppylang/hugr_builder/hugr.py:802]: Good fix there ;-). I wonder about `match n.op.root` ?

File: guppylang/prelude/_internal.py [hunk 24]
```diff
@@ -15,93 +15,120 @@
...
 
-class ConstIntS(BaseModel):
-    """Hugr representation of signed integers in the arithmetic extension."""
+class ConstInt(BaseModel):
+    """Hugr representation of signed and unsigned integers in the arithmetic extension.
+
+    Hugr always uses a u64 for the value. The interpretation is:
+      - as an unsigned integer, (value mod `2^N`);
+      - as a signed integer, (value mod `2^(N-1) - 2^(N-1)*a`)
+    where `N = 2^log_width` and `a` is the (N-1)th bit of `x` (counting from 0 = least
```
Comment (acl-cqc) [inline guppylang/prelude/_internal.py:53]: Quite detailed! I was going to suggest that it'd be easier just to say `'a' is the most significant bit` but I realize that there are bits beyond the N. It *might* still be easier just to start with "take the N least significant bits of the value. If the value is unsigned...."

Or even, "the interpretation is the twos-complement of the N least significant bits".

I mean, this is correct, so ok ;), but also quite hard to follow.

File: guppylang/tys/const.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply guppylang/tys/const.py:40]: Yeah. I was thinking this should be a Value, not the Const op-wrapper around one. Is this the thing that gets encoded as a TypeArg? Or used as a node in the Hugr? (If `ops.Const` is correct you might want to update the comment two lines above)

File: tests/error/iter_errors/end_missing.py (hunk 7 above)
Comment (acl-cqc) [reply tests/error/iter_errors/end_missing.py:9]: Ok, fair enough, this is clear enough. Certainly no need to intern everything! :)

Comment (acl-cqc) [reply tests/error/iter_errors/end_missing.py:9]: Ok, fair enough, this is clear enough. Certainly no need to intern everything! :)

File: tests/integration/test_poly.py (hunk 3 above)
Comment (acl-cqc) [reply tests/integration/test_poly.py:264]: I wonder if there is some syntax we could add in a future PR to let you manually specify those type args and get this example going if type inference doesn't cope and you really, really, want to. But how often does one do this, I don't think it's a priority!

Comment (acl-cqc) [reply tests/integration/test_poly.py:264]: I wonder if there is some syntax we could add in a future PR to let you manually specify those type args and get this example going if type inference doesn't cope and you really, really, want to. But how often does one do this, I don't think it's a priority!

**Review by acl-cqc — APPROVED** [review]
Already APPROVEd but APPROVE again :-)

TBH the thing here that has the worst "code smell" to my mind is DummyOp. But we *know* that that's a workaround for a not-yet-existing Hugr feature, so we can expect something a bit, ugh. But certainly, any *significant* further effort on this PR might be better directed towards that ;-)

File: guppylang/tys/const.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/tys/const.py:40]: Ohh sorry I misunderstood. Yes, `ops.Value` is better here

File: guppylang/prelude/_internal.py (hunk 24 above)
Comment (mark-koch) [reply guppylang/prelude/_internal.py:53]: This is copied straight from Hugr repo:

https://github.com/CQCL/hugr/blob/1c422abd7d3eb0538635fd8a307de80944f93bcf/hugr/src/std_extensions/arithmetic/int_types.rs#L78-L88
